src/insights_client/tests/test_metrics.py

```diff
@@ -285,3 +285,130 @@ def test_metrics_post_event_proxy(post, config_file_factory, rhsm_config_file_fa
         proxies={"https": "http://user:password@localhost:3128"},
     )
 
+@patch("insights_client.metrics.os.getenv")
+def test_is_offline(os_getenv, config_file_factory):
+    '''
+    Verify that _is_offline() produces correct output from the given configurations
+
+    NOTE: this test will become obsolete once InsightsConfig is used to load config
+    '''
+    # offline not specified
+    config_file = config_file_factory("")
+    cfg = configparser.RawConfigParser()
+    cfg.read(config_file.name)
+    os_getenv.return_value = None
+    with patch('sys.argv', ['insights-client']):
+        assert not _is_offline(cfg)
+    # offline specified in config
+    config_file = config_file_factory("offline=True\n")
+    cfg = configparser.RawConfigParser()
+    cfg.read(config_file.name)
+    os_getenv.return_value = None
+    with patch('sys.argv', ['insights-client']):
+        assert _is_offline(cfg)
+    # offline specified in CLI
+    config_file = config_file_factory("")
+    cfg = configparser.RawConfigParser()
+    cfg.read(config_file.name)
+    os_getenv.return_value = None
+    with patch('sys.argv', ['insights-client', '--offline']):
+        assert _is_offline(cfg)
+    # offline specified in env
+    config_file = config_file_factory("")
+    cfg = configparser.RawConfigParser()
+    cfg.read(config_file.name)
+    os_getenv.return_value = "True"
+    with patch('sys.argv', ['insights-client']):
+        assert _is_offline(cfg)
+
+@patch("insights_client.metrics._is_offline", Mock(return_value=True))
```

Comment:
Parametrization example.

```suggestion
@mark.parametrize(("config", "delenv", "setenv", "sysargv", "is_offline"), (
    ("", ("INSIGHTS_OFFLINE"), (), ["insights-client"], False),
    ("offline=True\n", ("INSIGHTS_OFFLINE"), (), ["insights-client"], True),
    ("", ("INSIGHTS_OFFLINE"), (), ["insights-client", "--offline"], True),
    ("", ("INSIGHTS_OFFLINE"), (("INSIGHTS_OFFLINE", "True"),), ["insights-client"], True),
))
def test_is_offline(monkeypatch, config, delenv, setenv, sysargv, is_offline, config_file_factory):
    '''
    Verify that _is_offline() produces correct output from the given configurations

    NOTE: this test will become obsolete once InsightsConfig is used to load config
    '''
    # offline not specified
    for d in delenv:
        monkeypatch.delenv(d, raising=False)
    for k, v in setenv:
        monkeypatch.setenv(k, v)
    monkeypatch.setattr("sys.argv", sysargv)

    config_file = config_file_factory(config)
    cfg = configparser.RawConfigParser()
    cfg.read(config_file.name)
    assert _is_offline(cfg) is is_offline

@patch("insights_client.metrics._is_offline", Mock(return_value=True))
```

Test split example.

```suggestion
def test_is_offline_default(monkeypatch, config_file_factory):
    config_file = config_file_factory("")
    cfg = configparser.RawConfigParser()
    cfg.read(config_file.name)
    monkeypatch.delenv("INSIGHTS_OFFLINE", raising=False)
    monkeypatch.setattr("sys.argv", ['insights-client'])
    assert not _is_offline(cfg)

def test_is_offline_config(monkeypatch, config_file_factory):
    config_file = config_file_factory("offline=True\n")
    cfg = configparser.RawConfigParser()
    cfg.read(config_file.name)
    monkeypatch.delenv("INSIGHTS_OFFLINE", raising=False)
    monkeypatch.setattr("sys.argv", ["insights-client"])
    assert _is_offline(cfg)

def test_is_offline_arg(monkeypatch, config_file_factory):
    config_file = config_file_factory("")
    cfg = configparser.RawConfigParser()
    cfg.read(config_file.name)
    monkeypatch.delenv("INSIGHTS_OFFLINE", raising=False)
    monkeypatch.setattr("sys.argv", ["insights-client", "--offline"])
    assert _is_offline(cfg)

def test_is_offline_env(monkeypatch, config_file_factory):
    config_file = config_file_factory("")
    cfg = configparser.RawConfigParser()
    cfg.read(config_file.name)
    monkeypatch.setenv("INSIGHTS_OFFLINE", "True")
    monkeypatch.setattr("sys.argv", ["insights-client"])
    assert _is_offline(cfg)

@patch("insights_client.metrics._is_offline", Mock(return_value=True))
```

I am more in favor of the parametrization, although it is a bit more complex, because then it would be simple to add more test cases. I mean those that would include overrides.

In one of the previous iterations I spotted an inconsistency (a bug) with the overrides, so it’s already happened. Also, the _strip_ and _lower_ calls as well as non-True/False strings are currently untested and with parametrization, adding those would become really easy.